src/rust/engine/src/scheduler.rs

```diff
@@ -112,46 +100,69 @@ impl Scheduler {
   }
 
   ///
-  /// Starting from existing roots, execute a graph to completion.
+  /// Attempts to complete a Node, recovering from `Failure::Invalidated` up to `count` times.
```

Comment:
Because the scope of a retry here is a Root, rather than an ExecutionRequest, I believe this gives rise to the possibility of inconsistent builds. It's possible my mental model is wrong, though.


I'm imagining, say, targets X and Y which both depend on Z, and asking for javadoc_transitive(X) and compile_transitive(Y). If javadoc_transitive(X) happens, then Z is modified, then compile_transitive(Y) happens, it will be told to retry, and compile_transitive(Y) will be built with post-modification Z, but javadoc_transitive(X) will be built with pre-modification Z.

It feels like it would be better to fail the whole build, or retry the whole build, rather than retrying individual roots. If there's no work to retry for a root unrelated to the modification, it will just hit the cache anyway.

Reply:
Yea, good call. Thanks for the scrutiny!